homeassistant/components/sensor/dsmr.py

```diff
@@ -0,0 +1,221 @@
+        """If result is an exception log it."""
+        _LOGGER.error('error during initialization of DSMR serial reader: %s',
+                      future.exception())
+    reader.add_done_callback(handle_error)
+
+    # add task to receive telegrams and update entities
+    hass.loop.create_task(dsmr.read_telegrams(queue))
+
+
+class DSMR:
+    """DSMR interface."""
+
+    def __init__(self, hass, config, devices):
+        """Setup DSMR serial interface, initialize, add device entity list."""
+        from dsmr_parser.serial import (
+            SERIAL_SETTINGS_V4,
+            SERIAL_SETTINGS_V2_2,
+            AsyncSerialReader
+        )
+        from dsmr_parser import telegram_specifications
+
+        # map dsmr version to settings
+        dsmr_versions = {
+            '4': (SERIAL_SETTINGS_V4, telegram_specifications.V4),
+            '2.2': (SERIAL_SETTINGS_V2_2, telegram_specifications.V2_2),
+        }
+
+        # initialize asynchronous telegram reader
+        dsmr_version = config[CONF_DSMR_VERSION]
+        self.dsmr_parser = AsyncSerialReader(
+            device=config[CONF_DEVICE],
+            serial_settings=dsmr_versions[dsmr_version][0],
+            telegram_specification=dsmr_versions[dsmr_version][1],
+        )
+
+        # keep list of device entities to update
+        self.devices = devices
+
+        # initialize empty telegram
+        self._telegram = {}
+
+    @asyncio.coroutine
+    def read_telegrams(self, queue):
+        """Receive parsed telegram from DSMR reader, update entities."""
+        while True:
+            # asynchronously get latest telegram when it arrives
+            self._telegram = yield from queue.get()
+            _LOGGER.debug('received DSMR telegram')
+
+            # make all device entities aware of new telegram
+            for device in self.devices:
+                yield from device.async_update_ha_state()
+
+    @property
+    def telegram(self):
+        """Return telegram object."""
+        return self._telegram
+
+
+class DSMREntity(Entity):
+    """Entity reading values from DSMR telegram."""
+
+    def __init__(self, name, obis, interface):
+        """"Initialize entity."""
+        # human readable name
+        self._name = name
+        # DSMR spec. value identifier
+        self._obis = obis
+        # interface class to get telegram data
+        self._interface = interface
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        if 'Power' in self._name:
+            return ICON_POWER
+        elif 'Gas' in self._name:
+            return ICON_GAS
+
+    @property
+    def state(self):
+        """Return the state of sensor, if available, translate if needed."""
+        from dsmr_parser import obis_references as obis
+
+        value = getattr(self._interface.telegram.get(self._obis, {}),
```

Comment:
Someone else also commented on this. I will refactor/document this piece a bit to make it more clear it is not  doing I/O.